Fixed Tab and Bookmark Right to Left Layout Drawing

ShaopengLin opened the pull request:
Improve #594 Fix #1142 

Changes:
- Tab text alignment in RTL mode fixed. It was showing the tail of the title instead of the start.
- The left border of tabs was missing in RTL. Now each tab has both a left and right border. This makes the border a little thicker than before, but better UX than no border.
- I forgot to add the border on both sides when I refactored the search bar, which meant it has no left border in RTL mode. Fixed
- Title text now also has 2px padding.

kelson42 reviewed (changes requested):
If you look to the icon, there is now a vertical alignement problem: I don't see why the two "Ted" Icon, and title should not be aligned!

ShaopengLin commented:
![Screenshot from 2024-07-30 01-41-25](https://github.com/user-attachments/assets/5da15b4f-2b3b-43b1-8263-847ea74ee7da)

@kelson42 The text is technically aligned to the icon. It is just the font size on the icon is different. If I misunderstood could you provide a screen shot of your alignment situation?

ShaopengLin commented:
Ignore my previous comment, It does seem the text is not aligning with respect to the icon...

kelson42 commented:
@ShaopengLin Please look at all this alignment/margin at the pixel level, so things look perfect.

ShaopengLin commented:
@kelson42 Additional Fixes:
1. Fixed the text to icon alignment.
2. The tab bottom border disappears when it is the current tab. The icon is up by 1px. Text painting now adjusts to this change as well.
3. Close button also does not adjust to the border disappearance. Fixed as well.

kelson42 commented:
Sorry, but this is still wrong:
![image](https://github.com/user-attachments/assets/0ba46ff1-2675-49f6-aa27-3d609be741e5)

* OK: What is green is OK, so the text is properly ligned verticaly
* OK: Top and bottom padding are the same at 5px, which is fine to me
* NOK: But why it left and right padding at 4px?! Should be 5 as well (or put 4pixel padding everywhere)
* NOK: Icon is 2 pixels too high!

ShaopengLin commented:
@kelson42 The weird behavior is from our padding and border handling. Now the padding should be 4px. The bottom padding is the same, we are just painting the border white as part of the focus animation.

![Screenshot from 2024-07-30 17-06-09](https://github.com/user-attachments/assets/a07c28a8-d8c7-4c3c-beab-e4a9a0d88c86)

kelson42 reviewed (approved):
LGTM
@veloman-yunkan Good for the code review

veloman-yunkan commented on src/tabbar.cpp:
```diff
@@ -463,7 +463,7 @@ void TabBar::paintEvent(QPaintEvent *e)
         int align = Qt::AlignVCenter;
         if (need_fade_out) {
-            align |= (right_to_left)? Qt::AlignRight : Qt::AlignLeft;
+            align |= (textRightToLeft == appRightToLeft) ? Qt::AlignLeft : Qt::AlignRight;
```
This looks strange. Why is the case of `textRightToLeft` and `appRightToLeft` both being `false` aligned the same way as when both of them are `true`?

ShaopengLin replied:
I will retest this and either make a explanation and visual, or correct it. A comment I believe will be needed here to explain the textbox behaviours.

ShaopengLin replied:
The alignment in RTL is flipped, i.e. if you set Align Left, it will actually align it to the right. https://doc.qt.io/qt-6/qtquick-positioning-righttoleft.html.

ShaopengLin commented:
@veloman-yunkan @kelson42  I just noticed that the padding code here isn't working as wanted in Qt6.2. I will make a change and then we can re-review.

ShaopengLin commented:
Done

veloman-yunkan commented on src/tabbar.cpp:
```diff
@@ -427,6 +427,16 @@ void TabBar::paintEvent(QPaintEvent *e)
 
         QRect tabTextRect = style()->subElementRect(QStyle::SE_TabBarTabText, &tabopt, this);
 
+#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
+        // Add Padding to left, right. Padding is 4px. Add 5 to account for 
+        // Extra pixel from border.
+        tabTextRect.setX(tabTextRect.x() + 5);
+        tabTextRect.setWidth(tabTextRect.width() - 5);
+#else
+        // Qt6 correctly adds left and right padding but now incorrectly adds
+        // top or bottom padding.
+        tabTextRect.setY(tabTextRect.y() - 4);
+#endif
```
Magic numbers are evil. Please introduce a local constant (`const int padding = 4)` with a comment linking it to the value of `QTabBar::tab::padding` in `resources/css/style.css` (also add a sibling comment in `resources/css/style.css`).

ShaopengLin replied:
Done

veloman-yunkan reviewed (approved):
LGTM

@kelson42 Please test again before merging
